### DISCUSSION TITLE
Why does pyenv shim a version of python it didn't install?

### DISCUSSION THREAD

**Discussion by AFlowOfCode** [pyenv/pyenv#2840] [Q&A] [answered]
```
$ which python3
/home/vagrant/.pyenv/shims/python3
$ pyenv which python3
/usr/bin/python3

$ python3
Python 3.8.10 (default, May 26 2023, 14:05:08)
[GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>>

$ pyenv uninstall 3.8.10
pyenv: version `3.8.10' not installed
$ pyenv uninstall 3.8
pyenv: version `3.8' not installed
$ pyenv uninstall 3
pyenv: version `3' not installed

$ ls -la /home/vagrant/.pyenv/versions
drwxrwxr-x  3 vagrant vagrant 4096 Jun  6  2022 .
drwxrwxr-x 14 vagrant vagrant 4096 Jun  6  2022 ..
drwxr-xr-x  6 vagrant vagrant 4096 Jun  6  2022 3.10.4
```

Can anybody explain how/why this might happen?

**Comment by native-api** [comment, marked as answer]
Pyenv creates shims for all executables that exist in Pyenv-managed installations.
Indeed, if you don't have any Pyenv-managed installations -- e.g. right after installaing Pyenv, or if you uninstall them all -- there would be no shims at all.

That's because Pyenv needs a facility to redirect them if any of the ways to select an alternative version is activated (the `PYENV_VERSION` envvar and `.python-version` files).
